### PR TITLE
project properties tabs, request result validation, project name editing, deletion of projects

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: backend/src/main/scala/com/griddynamics/genesis/repository/impl/ProjectPropertyRepository.scala [hunk 1]
```diff
@@ -24,27 +24,44 @@ package com.griddynamics.genesis.repository.impl
...
+          names.add(pp.name)
+          new model.ProjectProperty(projectId, pp.name, pp.value)
+        }
+      }
+    }
 
     GenesisSchema.projectProperties.deleteWhere(pp => pp.projectId === projectId)
     GenesisSchema.projectProperties.insert(modelProperties)
+
+    Option.empty
```
Comment (rsvato) [inline backend/src/main/scala/com/griddynamics/genesis/repository/impl/ProjectPropertyRepository.scala]: Did you mean None?

File: backend/src/main/scala/com/griddynamics/genesis/repository/impl/ProjectPropertyRepository.scala [hunk 2]
```diff
@@ -24,27 +24,44 @@ package com.griddynamics.genesis.repository.impl
...
   }
 
   @Transactional
-  def updateForProject(projectId: Int, properties : List[api.ProjectProperty]) {
-    val modelProperties = for (pp <- properties.map(convert(_))) yield new model.ProjectProperty(projectId, pp.name, pp.value)
+  def updateForProject(projectId: Int, properties : List[api.ProjectProperty]): Option[RequestResult] = {
+    val names = new HashSet[String](properties.size);
+    val modelProperties = for (pp <- properties.map(convert(_))) yield {
+      validate(pp) match {
+        case Some(result) => return Option.apply(result)
```
Comment (rsvato) [inline backend/src/main/scala/com/griddynamics/genesis/repository/impl/ProjectPropertyRepository.scala]: Some(result)

Option.apply is not meant to be called directly

File: backend/src/main/scala/com/griddynamics/genesis/repository/impl/ProjectPropertyRepository.scala [hunk 3]
```diff
@@ -24,27 +24,44 @@ package com.griddynamics.genesis.repository.impl
...
     val modelProperties = from(GenesisSchema.projectProperties)(pp => where(pp.projectId === projectId) select(pp) orderBy(pp.id asc)).toList
-
     modelProperties.map(convert(_))
   }
 
   @Transactional
-  def updateForProject(projectId: Int, properties : List[api.ProjectProperty]) {
-    val modelProperties = for (pp <- properties.map(convert(_))) yield new model.ProjectProperty(projectId, pp.name, pp.value)
+  def updateForProject(projectId: Int, properties : List[api.ProjectProperty]): Option[RequestResult] = {
+    val names = new HashSet[String](properties.size);
```
Comment (rsvato) [inline backend/src/main/scala/com/griddynamics/genesis/repository/impl/ProjectPropertyRepository.scala]: Why use java HashSet? Scala has mutable Set too.

After a second thought I really dislike this approach. Why not search for duplicates on list comprehension result?

File: backend/src/main/scala/com/griddynamics/genesis/repository/impl/ProjectPropertyRepository.scala [hunk 4]
```diff
@@ -56,4 +73,15 @@ class ProjectPropertyRepository extends AbstractGenericRepository[model.ProjectP
     projectProperty.id = dto.id
     projectProperty
   }
+
+  def validate(property: api.ProjectProperty): Option[RequestResult] = {
+    if ((property.name == null) || property.name.isEmpty) {
+      return Option.apply(new RequestResult(isSuccess = false, compoundServiceErrors = Seq("Project property name is empty")))
+    }
+    if (namePattern.findFirstIn(property.name) == None) {
```
Comment (rsvato) [inline backend/src/main/scala/com/griddynamics/genesis/repository/impl/ProjectPropertyRepository.scala:85]: None is not null. Don't use == to check for None

Comment (rsvato) [inline backend/src/main/scala/com/griddynamics/genesis/repository/impl/ProjectPropertyRepository.scala:85]: Scala has other mechanism to check for regular expression

File: backend/src/main/scala/com/griddynamics/genesis/repository/impl/ProjectPropertyRepository.scala [hunk 5]
```diff
@@ -24,27 +24,44 @@ package com.griddynamics.genesis.repository.impl
...
-
     modelProperties.map(convert(_))
   }
 
   @Transactional
-  def updateForProject(projectId: Int, properties : List[api.ProjectProperty]) {
-    val modelProperties = for (pp <- properties.map(convert(_))) yield new model.ProjectProperty(projectId, pp.name, pp.value)
+  def updateForProject(projectId: Int, properties : List[api.ProjectProperty]): Option[RequestResult] = {
+    val names = new HashSet[String](properties.size);
+    val modelProperties = for (pp <- properties.map(convert(_))) yield {
```
Comment (rsvato) [inline backend/src/main/scala/com/griddynamics/genesis/repository/impl/ProjectPropertyRepository.scala]: yield return?

File: frontend/src/main/scala/com/griddynamics/genesis/rest/ProjectPropertiesController.scala [hunk 6]
```diff
@@ -47,12 +45,16 @@ class ProjectPropertiesController extends RestApiExceptionsHandler {
...
-  def updateForProject(@PathVariable("projectId") projectId: Int, request: HttpServletRequest) {
+  def updateForProject(@PathVariable("projectId") projectId: Int, request: HttpServletRequest): RequestResult = {
     val properties = for (p <- GenesisRestController.extractParamsMapList(request)) yield {
       val name = GenesisRestController.extractValue("name", p)
       val value = GenesisRestController.extractValue("value", p)
       new api.ProjectProperty(0, 0, name, value)
     }
-    projectPropertyRepository.updateForProject(projectId, properties)
+
+    projectPropertyRepository.updateForProject(projectId, properties) match {
```
Comment (rsvato) [inline frontend/src/main/scala/com/griddynamics/genesis/rest/ProjectPropertiesController.scala]: Why do you need pattern matching here? Isn't better to make updateProject return a ready RequestResult with errors or isSuccess=true?

File: backend/src/main/scala/com/griddynamics/genesis/repository/impl/ProjectPropertyRepository.scala (hunk 3 above)
Comment (ghost) [inline backend/src/main/scala/com/griddynamics/genesis/repository/impl/ProjectPropertyRepository.scala]: What will be complexity of such search?
With Set complexity is O(1) for each element, and O(n) for full sequence.

File: backend/src/main/scala/com/griddynamics/genesis/repository/impl/ProjectPropertyRepository.scala (hunk 1 above)
Comment (ghost) [inline backend/src/main/scala/com/griddynamics/genesis/repository/impl/ProjectPropertyRepository.scala]: Yes, but it's not possible to return None directly, Idea showed me a bug.

**Pull request merged**